budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/persistence/contract/ContractRepository.java

```diff
@@ -20,56 +20,93 @@
      */
-    @Query("select new org.wickedsource.budgeteer.persistence.contract.ContractStatisticBean(:year+0," +
-            "case when (abs(cast(c.budget AS double)) < 10e-16) then null else (" +
-            "(SELECT coalesce(sum(wr.minutes * wr.dailyRate/ 60 / 8),0)" +
+    @Query("select new org.wickedsource.budgeteer.persistence.contract.ContractStatisticBean(:year+0," + //year
+            "case when (abs(cast(c.budget AS double)) < 10e-16) then null else (" + //progress
+            "((SELECT coalesce(sum(wr.minutes * wr.dailyRate/ 60 / 8),0)" +
             " FROM WorkRecordEntity wr" +
             " WHERE wr.budget.contract.id = :contractId AND (wr.year < :year OR (wr.year = :year AND wr.month <= :month)))" +
+            "+(select coalesce(sum(record.moneyAmount),0) from ManualRecordEntity record where record.budget.contract.id = :contractId and record.month = :month and record.year = :year))" +
             " / cast(c.budget AS double)" +
             ") end, " +
-            "(c.budget - coalesce((select sum(wr.minutes * wr.dailyRate/ 60 / 8) " +
+            "(c.budget - coalesce((select sum(wr.minutes * wr.dailyRate/ 60 / 8) " + //remaining
             "from WorkRecordEntity wr where wr.budget.contract.id = :contractId " +
             "AND (wr.year = :year AND wr.month = :month)" +
             "),0l)" +
+            "- (select coalesce(sum(record.moneyAmount),0) from ManualRecordEntity record where record.budget.contract.id = :contractId and record.month = :month and record.year = :year)"+
             ")," +
-            "coalesce((select sum(wr.minutes * wr.dailyRate/ 60 / 8)" +
+            "(coalesce((select sum(wr.minutes * wr.dailyRate/ 60 / 8)" + //spent
             "from WorkRecordEntity wr where wr.budget.contract.id = :contractId "+
             "AND (wr.year = :year AND wr.month = :month)" +
-            "),0l)," +
-            "coalesce((select sum(i.invoiceSum) from InvoiceEntity i where i.contract.id = :contractId AND (i.year = :year AND i.month = :month) ),0l)" +
+            "),0l)+ " +
+            "(select coalesce(sum(record.moneyAmount),0) from ManualRecordEntity record where record.budget.contract.id = :contractId and record.month = :month and record.year = :year))," +
+            "coalesce((select sum(i.invoiceSum) from InvoiceEntity i where i.contract.id = :contractId AND (i.year = :year AND i.month = :month) ),0l)" + //invoiced
             ",:month +0" +
             ") from ContractEntity c where c.id = :contractId")
     ContractStatisticBean getContractStatisticByMonthAndYear(@Param("contractId") Long contractId, @Param("month") Integer month, @Param("year") Integer year);
 
-    @Query("select coalesce(sum(wr.minutes * wr.dailyRate/ 60 / 8),0) from WorkRecordEntity wr where wr.budget.contract.id = :contractId")
+    @Query("select (coalesce(sum(wr.minutes * wr.dailyRate/ 60 / 8),0) +" +
+           " (select coalesce(sum(record.moneyAmount),0) " +
```

Comment:
This query doesn't work with oracle database!
So it's currently not possible to create new contracts, if you use a oracle database.

Reply:
This is what hibernate generates off the statement annotated on `getBudgetLeftByContractId` (pretty print by hand):

```sql
select nvl(sum(workrecord0_.minutes * workrecord0_.daily_rate / 60 / 8), 0) + ( 
    select nvl(sum(manualreco1_.money_amount), 0) 
    from manual_record_entity manualreco1_, budget budgetenti2_ 
    where manualreco1_.budget_id=budgetenti2_.id and 
        budgetenti2_.contract_id=?) as col_0_0_ 
from work_record workrecord0_ 
    cross join budget budgetenti3_ 
where workrecord0_.budget_id=budgetenti3_.id and 
    budgetenti3_.contract_id=?
```

The addition of seemingly independent information sets is thus integrated in a subselect clause where the inner one (select from manual_record and budget) may (but does not) depend on the outer query. In order for it to depend on the outer query in would however need to group the information by budget - note that the inner select happens on the cross product (any combination) of manual record and budget applicable to their join condition, and each thus combination would be applied to the outer select. This is simply not necessary, and in SQL this can be reformulated as such:

```sql
select nvl(sum(workrecord0_.minutes*workrecord0_.daily_rate/60/8), 0) + 
    nvl(sum(manualreco1_.money_amount), 0)
from work_record workrecord0_ 
cross join budget budgetenti3_ 
cross join manual_record_entity manualreco1_
where workrecord0_.budget_id=budgetenti3_.id and 
    manualreco1_.budget_id=budgetenti3_.id and
    budgetenti3_.contract_id=:contractId;
```

I'm currently not quite sure how this could be expressed in HQL so unless you wish to embed native SQL queries I suggest to split the statement in two more operable - I do not see any recognizable performance loss and it would hugely improve upon maintainability.

Reply:
Note that `Double getBudgetLeftByContractId(@Param("contractId") long contractId);` would not work as well due to the same problem.

Reply:
I couln't find a good  and working way to change the queries, so I split them up.
See my comment in #339 .